src/services/connection-pool.ts

```diff
@@ -0,0 +1,68 @@
+import * as mysql2 from "mysql2/promise";
+import {Connection, FieldPacket, Pool} from "mysql2/promise";
+import {Maybe} from "../models/optionals";
+import {getConnectionPoolOptions} from "./connection-pool-options";
+
+export interface QueryResponse {
+    rows?: any;
+    fields?: FieldPacket[];
+}
+
+// Lazy
+let pool: Maybe<Pool>;
+
+export const getConnectionPool = async (): Promise<Pool> => {
+    if (!pool) {
+        console.info("getConnectionPool: Creating connection pool...");
+
+        const config = await getConnectionPoolOptions();
+
+        console.info(`getConnectionPool: Destination DB ${config.database}`);
+
+        pool = mysql2.createPool(config);
+
+        console.info("getConnectionPool: Connection pool created");
+    }
+    return pool;
+};
+
+export const destroyConnectionPool = async (): Promise<void> => {
+    console.info("destroyConnectionPool: Destroying connection pool...");
+    if (pool) {
+        await pool.end();
+        pool = undefined;
+        console.info(`destroyConnectionPool: connection pool destroyed`);
+    } else {
+        console.info(`destroyConnectionPool: nothing to do`);
+    }
+};
+
+export const executeSql = async (sql: string, templateVariables?: any[], connection?: Connection): Promise<QueryResponse> => {
+    if (templateVariables) {
+        templateVariables = undefinedToNull(templateVariables);
+    }
+
+    console.info(`Executing SQL: ${sql}`);
+    console.info(`Template vars: ${templateVariables === undefined ? "[]" : templateVariables}`);
+
+    if (connection) {
+        const [rows, fields] = await connection.execute(sql, templateVariables);
+        return {rows, fields};
+    } else {
+        const connectionPool = await getConnectionPool();
+        const [rows, fields] = await connectionPool.execute(sql, templateVariables);
+        return {rows, fields};
+    }
+};
+
+// npm packages mysql and mysql2 will throw an error on encountering JS "undefined".
+// In contrast, an explicit JS "null" translates to SQL "NULL".
+// hence, transform all "undefined" to "null".
+const undefinedToNull = (array: any[]): any[] => {
```

Comment:
It is probably one for all of us and is happening because of the joy of developing with Js and perhaps unnecessary since the data seems to be flowing one way - dynamo data -> sql data which is then used for query purposes by analysts?
By transforming undefined to null we are losing its initial intent so i was thinking sql would allow undefined to be referenced as a string ? I will also need some clarity around what type of queries will be performed and use cases for people consuming the service to understand better?

Reply:
Defining `undefined` values as the raw string `undefined` is a decision for the business intelligence department / the analysts who will actually be using this SQL. If it was me, personally, I'd want both explicit `null` and missing values to translate to SQL `NULL`. @SanjeetBJSS can I get some thoughts on this?

There is also the question of existing CVS API specs, which have two problems related to nulls:
* Some fields are required, but also nullable. Passing anything but an explicit null on these fields causes a `400 Bad Request`.
* Some fields are optional, and nullable, but passing an explicit null to them causes a `400 Bad Request`.

This is unreasonable and extremely hard to follow / implement in code (see my [`ExplicitNullAdapterFactory`](https://github.com/dvsa/cvs-auto-svc-vott/blob/feature/CVSB-19155/src/main/java/vott/models/adapter/ExplicitNullAdapterFactory.java) helping out the regression guys). I wanted to follow a simpler approach for the SQL side of things

Reply:
Hi Guys, I would like undefined set to NULL. The fact it is a string, is easily established by the metadata data, remember this is RDS now.